packages/dds/tree/src/codec/codec.ts

```diff
@@ -0,0 +1,177 @@
+	 */
+	resolve(formatVersion: number): IMultiFormatCodec<TDecoded>;
+
+	/**
+	 * @returns - an iterable of all format versions supported by this family.
+	 */
+	getSupportedFormats(): Iterable<number>;
+}
+
+/**
+ * Creates a codec family from a registry of codecs.
+ * Any codec that is not a {@link IMultiFormatCodec} will be wrapped with a default binary encoding.
+ */
+export function makeCodecFamily<TDecoded>(
+	registry: Iterable<
+		[formatVersion: number, codec: IMultiFormatCodec<TDecoded> | IJsonCodec<TDecoded>]
+	>,
+): ICodecFamily<TDecoded> {
+	const isJsonCodec = (
+		codec: IMultiFormatCodec<TDecoded> | IJsonCodec<TDecoded>,
+	): codec is IJsonCodec<TDecoded> =>
+		typeof (codec as any).encode === "function" && typeof (codec as any).decode === "function";
+	const codecs: Map<number, IMultiFormatCodec<TDecoded>> = new Map();
+	for (const [formatVersion, codec] of registry) {
+		if (codecs.get(formatVersion) !== undefined) {
+			fail("Duplicate codecs specified.");
+		}
+		if (isJsonCodec(codec)) {
+			codecs.set(formatVersion, withDefaultBinaryEncoding(codec));
+		} else {
+			codecs.set(formatVersion, codec);
+		}
+	}
+
+	return {
+		resolve(formatVersion: number) {
+			const codec = codecs.get(formatVersion);
+			assert(codec !== undefined, "Requested coded for unsupported format.");
+			return codec;
+		},
+		getSupportedFormats() {
+			return codecs.keys();
+		},
+	};
+}
+
+class DefaultBinaryCodec<TDecoded> implements IBinaryCodec<TDecoded> {
+	public constructor(private readonly jsonCodec: IJsonCodec<TDecoded>) {}
+
+	public encode(change: TDecoded): IsoBuffer {
+		const jsonable = this.jsonCodec.encode(change);
+		const json = JSON.stringify(jsonable);
+		return IsoBuffer.from(json);
+	}
+
+	public decode(change: IsoBuffer): TDecoded {
+		const json = bufferToString(change, "utf8");
+		const jsonable = JSON.parse(json);
+		return this.jsonCodec.decode(jsonable);
+	}
+}
+
+/**
+ * Constructs a {@link IMultiFormatCodec} from a `IJsonCodec` using a generic binary encoding that simply writes
+ * the json representation of the object to a buffer.
+ */
+export function withDefaultBinaryEncoding<TDecoded>(
+	jsonCodec: IJsonCodec<TDecoded>,
+): IMultiFormatCodec<TDecoded> {
+	return {
+		json: jsonCodec,
+		binary: new DefaultBinaryCodec(jsonCodec),
+	};
+}
+
+/**
+ * Codec for objects which carry no information.
+ */
+export const unitCodec: IMultiFormatCodec<0> = {
+	json: {
+		encode: () => 0,
+		decode: () => 0,
+	},
+	binary: {
+		encode: () => IsoBuffer.from(""),
+		decode: () => 0,
+	},
+};
+
+/**
+ * Creates a json codec for objects which are just a json compatible value
+ * and can be serialized as-is.
+ *
+ * @remarks - Beware that this encoder doesn't validate its input and isn't typesafe.
+ * It would be great to be able to constrain T to a reasonable type, but due to how typechecking
+ * of index signatures works, JsonCompatibleReadOnly isn't sufficient.
+ */
+export function makeValueCodec<T>(): IJsonCodec<T> {
```

Comment:
this is the one edge case where Jsonable is useful. Its not fully safe, but it can help here.

Reply:
actually, nevermind, it only helps when you can infer the type paramater. I misread what this is doing.